finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+        self.model_config = get_config_from_model_name(self.args.model_name)
+    
+    def get_memory_statistics(self, precision: int = 3) -> Dict[str, Any]:
+        memory_allocated = None
+        memory_reserved = None
+        max_memory_allocated = None
+        max_memory_reserved = None
+        
+        if torch.cuda.is_available():
+            device = torch.cuda.current_device()
+            memory_allocated = torch.cuda.memory_allocated(device)
+            memory_reserved = torch.cuda.memory_reserved(device)
+            max_memory_allocated = torch.cuda.max_memory_allocated(device)
+            max_memory_reserved = torch.cuda.max_memory_reserved(device)
+        
+        elif torch.mps.is_available():
+            memory_allocated = torch.mps.current_allocated_memory()
+        
+        else:
+            logger.warning("No CUDA, MPS, or ROCm device found. Memory statistics are not available.")
+        
+        return {
+            "memory_allocated": round(bytes_to_gigabytes(memory_allocated), ndigits=precision),
+            "memory_reserved": round(bytes_to_gigabytes(memory_reserved), ndigits=precision),
+            "max_memory_allocated": round(bytes_to_gigabytes(max_memory_allocated), ndigits=precision),
+            "max_memory_reserved": round(bytes_to_gigabytes(max_memory_reserved), ndigits=precision),
+        }
+
+    def prepare_models(self) -> None:
+        logger.info("Initializing models")
+
+        # TODO(aryan): refactor in future
+        load_components_kwargs = {
+            "text_encoder_dtype": torch.bfloat16,
+            "transformer_dtype": torch.bfloat16,
+            "vae_dtype": torch.bfloat16,
+            "cache_dir": self.args.cache_dir,
+        }
+        if self.args.pretrained_model_name_or_path is not None:
+            load_components_kwargs["model_id"] = self.args.pretrained_model_name_or_path
+        components = self._model_config_call(self.model_config["load_components"], load_components_kwargs)
+
+        self.tokenizer = components.get("tokenizer", None)
+        self.text_encoder = components.get("text_encoder", None)
+        self.transformer = components.get("transformer", None)
+        self.vae = components.get("vae", None)
+        self.scheduler = components.get("scheduler", None)
+
+    def prepare_dataset(self) -> None:
+        logger.info("Initializing dataset and dataloader")
+
+        self.dataset = VideoDatasetWithResizing(
+            data_root=self.args.data_root,
+            caption_column=self.args.caption_column,
+            video_column=self.args.video_column,
+            resolution_buckets=self.args.video_resolution_buckets,
+            dataset_file=self.args.dataset_file,
+            id_token=self.args.id_token,
+        )
+        self.dataloader = torch.utils.data.DataLoader(
+            self.dataset,
+            batch_size=1,
+            sampler=BucketSampler(self.dataset, batch_size=self.args.batch_size, shuffle=True),
```

Comment:
Bucket sampling could be made configurable.

Reply:
For now, I will keep it as is and work on the abstractions for this later